### DISCUSSION TITLE
window is undefined

### DISCUSSION THREAD

**Discussion by naissa12** [vercel/next.js#11156] [Help] [answered]
I have the following page
```

if (document.cookie) {
   loginLink = "xxxx";
}

export default () => (
    <div className="header" id="header">
        <a href="/" className="logo"></a>
        <a href="/books">Explore</a>
        <a href="/search">Search</a>
        <a href="/pricing">Upgrade to Premium</a>
        <a href="/teams">Teams</a>
        <a href="/blog">Blog</a>
        {loginLink}
        <a href="#" className="icon" onClick={toggleMobile.bind(this)}>
            <div className="line"></div>
            <div className="line"></div>
            <div className="line"></div>
        </a>
        <div className="clear"></div>
    </div>
)

```

I get an error that document.cookie when creating Next.js pages. Same for window.

**Comment by lfades** [comment]
That's expected because Next.js is server rendered by default, we do it this way to statically generate HTML at build time and for Server Side Rendering. You can read more about that [in the documentation](https://nextjs.org/docs/basic-features/pages#pre-rendering).

There's no fix for this, instead what you should do is to only access browser APIs after the first render (aka after hydration), usually done inside [`useEffect`](https://reactjs.org/docs/hooks-effect.html).

In other words, you can assume that there's no browser instance for the first render, because it happened in the server at build time (SSG), or at runtime (SSR).

**Comment by timneutkens** [comment, marked as answer]
As @lfades explained you shouldn't access browser environment related things in the global scope. They should be accessed using the `useEffect` hook:

```
import {useState, useEffect} from 'react'

export default () => {
  const [hasCookie, setHasCookie] = useState(false)
  useEffect(() => {
    if(document.cookie) {
      setHasCookie(true)
    }
  }, [])

  return (
    <div className="header" id="header">
        <a href="/" className="logo"></a>
        <a href="/books">Explore</a>
        <a href="/search">Search</a>
        <a href="/pricing">Upgrade to Premium</a>
        <a href="/teams">Teams</a>
        <a href="/blog">Blog</a>
        {hasCookie && <a href="/dashboard">Dashboard</a>}
        <a href="#" className="icon" onClick={toggleMobile.bind(this)}>
            <div className="line"></div>
            <div className="line"></div>
            <div className="line"></div>
        </a>
        <div className="clear"></div>
    </div>
  )
}
```

**Reply by flutterforsale** [reply]
Hi @timneutkens 

In this case:  use `useEffect` what is the difference with use `if (process.browser){}`

**Reply by timneutkens** [reply]
> In this case: use useEffect what is the difference with use if (process.browser){}

hydration errors causing the whole tree to be thrown away.

**Reply by vishalshah6153** [reply]
@timneutkens I'm facing the same issue. I'm trying to get the width with the window.innerLength. And I tried useEffect too. But the window object stays undefined. Any idea why is that?

**Comment by yaohuangguan** [comment]
try doing your function in if statement like, typeof window !== 'undefined', then....

**Reply by timneutkens** [reply]
That's not the right way to do it and will break hydration. See https://github.com/zeit/next.js/discussions/11156#discussioncomment-1573

**Comment by phoenisx** [comment]
As this discussion seems relevant for my problem, posting my doubt here.

I understand NextJS does hydration on  compile time, so does that mean to use any Browser API in Static pages, pages have to be rendered one extra time every time for first page load?? 

I say that, because I am experiencing same behavior in my static pages. I am trying to use `window.localStorage` to get my theme color, and if it is set to something other than default value, page first loads with default theme and then re-renders with the user saved theme color stored in `localStorage`.

Is there no way to handle such extra first time re-renders, just for Static pages??
